pkg/asset/installconfig/openstack/session.go

```diff
@@ -21,6 +22,12 @@ type Session struct {
 // GetSession returns an OpenStack session for a given cloud name in clouds.yaml.
 func GetSession(cloudName string) (*Session, error) {
 	opts := defaultClientOpts(cloudName)
+
+	// We should unset OS_CLOUD env variable here, because the real cloud name was
+	// defined on the previous step. OS_CLOUD has more priority, so the value from
+	// "opts" variable will be ignored if OS_CLOUD contains something.
+	os.Unsetenv("OS_CLOUD")
```

Comment:
what! why is the library taking env as priority over what's actually being provided to it explicitly?

Also I think you will have to remember to make this change everytime you use the library function, and there is high chance for forgetting this. Can you modify the invocation of the GetCloudFromYAML such that it either changes the order of priority or stop using the env.

unsetting the env is no really a good behavior, it is racey and can have unintended effect on the user environement.

Reply:
The gophercloud code that confirms this behavior: https://github.com/openshift/installer/blob/aaff33ec09d42d527a424a723e144449382c6e9e/vendor/github.com/gophercloud/utils/openstack/clientconfig/requests.go#L717-L733

Reply:
As you can notice, "cloud" overwriting happens before the invocation of `GetCloudFromYAML`, so its modification won't help here.

Reply:
But the function you linked above is not something you are actually invoking here.. [NewServiceClient](https://github.com/openshift/installer/blob/aaff33ec09d42d527a424a723e144449382c6e9e/vendor/github.com/gophercloud/utils/openstack/clientconfig/requests.go#L709) and you are directly calling GetCloudFromYAML next after the unset.

So why are we unsetting here again?

Reply:
looking at the function for region https://github.com/openshift/installer/blob/aaff33ec09d42d527a424a723e144449382c6e9e/vendor/github.com/gophercloud/utils/openstack/clientconfig/requests.go#L756-L772

here the precedence order is opt > env . which is same as endpoint but different from the cloud order?

Reply:
GetSession is called in several places in the code (to generate tfvars, for instance), so I decided to unset the variable here to reduce duplication of os.Unsetenv("OS_CLOUD") later. 

Also please notice, that Terraform's OpenStack Provider also uses gophercloud library, and this behavior is maintained there as well.

Reply:
> Also please notice, that Terraform's OpenStack Provider also uses gophercloud library, and this behavior is maintained there as well.

can you link to code that unsets the env. also can we open an issue with gophercloud to fix their order, this currently is just odd imo.

Reply:
I looked at the code, they've already created a separate auth implementation in the same repo: https://github.com/gophercloud/utils/tree/master/terraform/auth
So, Terraform is safe.